### ISSUE TITLE
[pigeon]plugin: java.lang.NullPointerException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkNotNullParameter, 

### ISSUE THREAD

**Issue by hello-soon** [flutter/flutter#139194]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

 **pigeon: 14.0.0**

<img width="652" alt="image" src="https://github.com/flutter/flutter/assets/62131847/158fed3b-9969-4384-9d37-f8ecf7c8af5b">

but!

<img width="974" alt="image" src="https://github.com/flutter/flutter/assets/62131847/5d5faa77-cd7b-4615-b520-89e318f69609">

### Expected results

 Parameter specified as non-null is null

### Actual results

 Parameter specified as non-null is null

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by danagbemava-nc** [comment]
Hi @hello-soon, thanks for filing this issue 

Please provide a minimal reproducible code sample and steps to reproduce so that we can investigate this issue. 

Thank you

**Comment by hello-soon** [comment]
#### use `checkReport` in kotlin,    can not  implement `Result` with nullable `result`
<img width="845" alt="image" src="https://github.com/flutter/flutter/assets/62131847/b46770c8-59ac-417a-b25e-cbf64762f156">

#### must be :
<img width="800" alt="image" src="https://github.com/flutter/flutter/assets/62131847/800a6695-f01a-4055-819a-946361e6645c">

when result is null, it throw NullPointerException


#### solution:
Do not use it in kotlin, you need to use it in java
<img width="782" alt="image" src="https://github.com/flutter/flutter/assets/62131847/73613f8d-8f6d-4b9f-bb1a-52bca4fde8d1">

**Comment by MeandNi** [comment]
same issue when FlutterApi method return void @danagbemava-nc 

```dart
@FlutterApi()
abstract class XXXX {
  void xxx();
}
```

**Comment by danagbemava-nc** [comment]
HI @hello-soon, @MeandNi, can either of you please provide the pigeon configuration that you used to generate the code you're having the issue with? 

Thank you

**Comment by ben-xD** [comment]
It's affecting me as well. I have a function (that has worked in the past, with older pigeon versions: v10 was what I used), but upgrading causes this issue.

The function is a `@FlutterApi` called from the Android side, into the Dart side. In my case, it's telling Flutter there's a new push notification:

### API
```dart
@FlutterApi()
abstract class PushFlutterApi {
  void onBackgroundMessage(RemoteMessage message);
}
```

The pigeon-generated code has `result.success(null)`, but then it throws an error: `Parameter specified as non-null is null`.

## Problem: Generated Java code contradicts itself

I guess the generated function `onBackgroundMessage` should accept a `NullableResult`, not a `Result`, since my function doesn't return anything. Currently, it's asking for a `Result`, and so I give it that, but `Result` **validates stuff to be non null** so it crashes. If I give it what it actually needs (and changing just the function signature), it works.

![contra](https://github.com/flutter/flutter/assets/24711048/d2473076-6162-4c39-88ce-5c31a9cf2f15)

```java
    public void onBackgroundMessage(@NonNull RemoteMessage messageArg, @NonNull Result<Void> result) {
      final String channelName = "dev.flutter.pigeon.push_platform_interface.PushFlutterApi.onBackgroundMessage";
      BasicMessageChannel<Object> channel =
          new BasicMessageChannel<>(
              binaryMessenger, channelName, getCodec());
      channel.send(
          new ArrayList<Object>(Collections.singletonList(messageArg)),
          channelReply -> {
            if (channelReply instanceof List) {
              List<Object> listReply = (List<Object>) channelReply;
              if (listReply.size() > 1) {
                result.error(new FlutterError((String) listReply.get(0), (String) listReply.get(1), (String) listReply.get(2)));
              } else {
                result.success(null);
              }
            }  else {
              result.error(createConnectionError(channelName));
            } 
          });
    }
```

```java
  /** Asynchronous error handling return type for non-nullable API method returns. */
  public interface Result<T> {
    /** Success case callback method for handling returns. */
    void success(@NonNull T result);

    /** Failure case callback method for handling errors. */
    void error(@NonNull Throwable error);
  }
```

## Temporary solution

TLDR: My solution is to make changes to the generated pigeon code, just to use `NullableResult` instead

![using NullableResult](https://github.com/flutter/flutter/assets/24711048/8af48de2-3001-4f99-8f6b-3bb9ff183610)

**Comment by danagbemava-nc** [comment]
Hi @ben-xD, thanks for info. 
I see the same thing when the code is generated. 

Labeling for further investigation. 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.16.5, on macOS 14.1.2 23B92 darwin-arm64, locale en-GB)
    • Flutter version 3.16.5 on channel stable at /Users/nexus/dev/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (9 days ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.3.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (3 available)
    • Nexus (mobile)  • 00008020-001875E83A38002E • ios            • iOS 17.2.1 21C66
    • macOS (desktop) • macos                     • darwin-arm64   • macOS 14.1.2 23B92 darwin-arm64
    • Chrome (web)    • chrome                    • web-javascript • Google Chrome 120.0.6099.129
    ! Error: Browsing on the local area network for Dean’s iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```
  
```
[!] Flutter (Channel master, 3.18.0-18.0.pre.44, on macOS 14.1.2 23B92 darwin-arm64, locale en-GB)
    • Flutter version 3.18.0-18.0.pre.44 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 420f18be76 (2 hours ago), 2023-12-29 13:45:36 +0800
    • Engine revision 7a7ddc53cd
    • Dart version 3.3.0 (build 3.3.0-270.0.dev)
    • DevTools version 2.31.0-dev.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.3.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (3 available)
    • Nexus (mobile)  • 00008020-001875E83A38002E • ios            • iOS 17.2.1 21C66
    • macOS (desktop) • macos                     • darwin-arm64   • macOS 14.1.2 23B92 darwin-arm64
    • Chrome (web)    • chrome                    • web-javascript • Google Chrome 120.0.6099.129
    ! Error: Browsing on the local area network for Dean’s iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by stuartmorgan** [comment]
`Void` is a weird type in Java; it can't be instantiated, so the only thing we *can* pass to it is `null`. Given that, it looks like we probably need to special-case void FlutterApi functions to either:
- use `NullableResult` instead, or
- introduce a new `VoidResult` class which just uses `success()` for the signature

Since we have to special case it either way, I would lean toward the latter.

(Out of curiosity, if you are interoperating with Kotlin code, is there a reason you are using the Java generator rather than the Kotlin generator?)

**Comment by ben-xD** [comment]
Nice, I've tested pigeon 16.0.0 and it works. Problem solved, thanks
